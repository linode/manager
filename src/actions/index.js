export { setTitle } from './title';
export { setSource } from './source';
export { setAnalytics } from './analytics';
