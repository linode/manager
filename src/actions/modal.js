import { EmitEvent } from 'linode-components';

export const SHOW_MODAL = '@@modal/SHOW_MODAL';
export const HIDE_MODAL = '@@modal/HIDE_MODAL';

export function showModal(title) {
  EmitEvent('modal:show', 'Modal', 'show');
  return { type: SHOW_MODAL, title };
}

export function hideModal() {
  return { type: HIDE_MODAL };
}
