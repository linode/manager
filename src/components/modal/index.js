export { default as PortalModal } from './PortalModal';
