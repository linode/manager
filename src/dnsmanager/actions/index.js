export const TOGGLE_SELECTED = '@@zones@@index/TOGGLE_SELECTED';

export function toggleSelected(id) {
  return { type: TOGGLE_SELECTED, selected: [id] };
}

export function toggleSelectAll() {
  return async (dispatch, getState) => {
    const zones = Object.keys(getState().api.dnszones.dnszones);
    dispatch({ type: TOGGLE_SELECTED, selected: zones });
  };
}
