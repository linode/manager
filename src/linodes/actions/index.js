export const TOGGLE_SELECTED = '@@linodes@@index/TOGGLE_SELECTED';

export function toggleSelected(id) {
  const selected = Array.isArray(id) ? id : [id];
  return { type: TOGGLE_SELECTED, selected };
}

// TODO: may be appropriate to remove, depending on filtering/pagination
export function toggleSelectAll() {
  return async (dispatch, getState) => {
    const ids = getState().api.linodes.ids;
    dispatch({ type: TOGGLE_SELECTED, selected: ids });
  };
}
