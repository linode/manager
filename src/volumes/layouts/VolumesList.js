import omitBy from 'lodash/omitBy';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PrimaryButton from 'linode-components/dist/buttons/PrimaryButton';

import { setAnalytics, setSource } from '~/actions';
import api from '~/api';
import ChainedDocumentTitle from '~/components/ChainedDocumentTitle';
import CreateHelper from '~/components/CreateHelper';

import { AddEditVolume, VolumesList as VolumesListComponent } from '../components';
import { ComponentPreload as Preload } from '~/decorators/Preload';


const OBJECT_TYPE = 'volumes';

export class VolumesList extends Component {
  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));
    dispatch(setAnalytics(['volumes']));
  }

  render() {
    const { dispatch, linodes, volumes: allVolumes, selectedMap } = this.props;

    // Hack because the API is currently returning deleted volumes.
    const volumes = omitBy(allVolumes.volumes, v => v.status === 'deleted');

    return (
      <div className="PrimaryPage container">
        <ChainedDocumentTitle title="Volumes" />
        <header className="PrimaryPage-header">
          <div className="PrimaryPage-headerRow clearfix">
            <h1 className="float-sm-left">Volumes</h1>
            <PrimaryButton
              className="float-sm-right"
              onClick={() => AddEditVolume.trigger(dispatch, linodes)}
            >
              Add a Volume
            </PrimaryButton>
          </div>
        </header>
        <div className="PrimaryPage-body">
          {Object.keys(volumes).length ?
            <VolumesListComponent
              objectType={OBJECT_TYPE}
              volumes={volumes}
              linodes={linodes.linodes}
              selectedMap={selectedMap}
              dispatch={dispatch}
              linodes={linodes}
            /> :
            <CreateHelper
              label="Volumes"
              onClick={() => AddEditVolume.trigger(dispatch, linodes)}
              linkText="Add a Volume"
            />
          }
        </div>
      </div>
    );
  }
}

VolumesList.propTypes = {
  dispatch: PropTypes.func,
  volumes: PropTypes.object,
  linodes: PropTypes.object,
  selectedMap: PropTypes.object.isRequired,
};


function mapStateToProps(state) {
  return {
    volumes: state.api.volumes,
    linodes: state.api.linodes.linodes,
    selectedMap: state.select.selected[OBJECT_TYPE] || {},
  };
}

const preloadRequest = async (dispatch) => {
  await Promise.all([
    api.linodes, api.volumes,
  ].map(o => dispatch(o.all())));
};

export default compose(
  connect(mapStateToProps),
  Preload(preloadRequest)
)(VolumesList);
