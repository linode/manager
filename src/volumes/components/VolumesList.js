import filter from 'lodash/filter';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import LinkButton from 'linode-components/dist/buttons/LinkButton';
import Dropdown from 'linode-components/dist/dropdowns/Dropdown';
import Input from 'linode-components/dist/forms/Input';
import List from 'linode-components/dist/lists/List';
import ListBody from 'linode-components/dist/lists/bodies/ListBody';
import ListHeader from 'linode-components/dist/lists/headers/ListHeader';
import MassEditControl from 'linode-components/dist/lists/controls/MassEditControl';
import Table from 'linode-components/dist/tables/Table';
import CheckboxCell from 'linode-components/dist/tables/cells/CheckboxCell';
import LabelCell from 'linode-components/dist/tables/cells/LabelCell';
import TableCell from 'linode-components/dist/tables/cells/TableCell';

import { default as toggleSelected } from '~/actions/select';
import api from '~/api';
import { actions as linodeActions } from '~/api/generic/linodes';
import { detachVolume } from '~/api/ad-hoc/volumes';
import { transform } from '~/api/util';
import { RegionCell } from '~/components/tables/cells';
import { confirmThenDelete } from '~/utilities';

import AddEditVolume from './AddEditVolume';
import MoreInfo from './MoreInfo';
import AttachVolume from './AttachVolume';
import ResizeVolume from './ResizeVolume';


export default class VolumesList extends Component {
  constructor(props) {
    super(props);

    this.state = { filter: '' };
  }

  detachVolumes = confirmThenDelete(
    this.props.dispatch,
    'volume',
    this.removeFromLinodeAndCall(detachVolume),
    this.props.objectType,
    undefined,
    'detach',
    'detaching'
  ).bind(this);


  deleteVolumes = confirmThenDelete(
    this.props.dispatch,
    'volume',
    this.removeFromLinodeAndCall(api.volumes.delete),
    this.props.objectType
  ).bind(this);

  removeFromLinodeAndCall(action) {
    return id => async (dispatch, getState) => {
      let volumeOnLinode = getState().api.volumes.volumes[id];
      await dispatch(action(id));

      try {
        let linodeId;

        if (!volumeOnLinode) {
          Object.values(getState().api.linodes.linodes).forEach(l => {
            const [volume] = filter(Object.values(l._volumes.volumes), v => v.id === id);
            if (volume) {
              volumeOnLinode = volume;
              linodeId = volume.linode_id;
            }
          });
        } else {
          linodeId = volumeOnLinode.linode_id;
        }

        if (linodeId) {
          return dispatch(linodeActions.volumes.delete(linodeId, volumeOnLinode.id));
        }
      } catch (e) {
        // Pass
      }
    };
  }

  renderAttached = ({ column, record }) => {
    const { linodes } = this.props;
    const { linode_id: linodeId } = record;

    let contents = <span>Unattached</span>;
    let to = linodeId;

    if (to) {
      const linode = linodes[linodeId];
      if (linode && linode.label) {
        to = (
          <LinkButton to={`/linodes/${linode.label}/settings/advanced`}>
            {linode.label}
          </LinkButton>
        );
      }

      contents = <div>Attached to {to}</div>;
    }

    return (
      <TableCell column={column} record={record}>
        {contents}
      </TableCell>
    );
  }

  renderVolumeActions = ({ column, record }) => {
    const { dispatch, linodes } = this.props;

    const groups = [
      { elements: [{ name: 'More Info', action: () => MoreInfo.trigger(dispatch, record) }] },
      {
        elements: [
          {
            name: 'Edit',
            action: () => AddEditVolume.trigger(dispatch, linodes, record),
          },
          {
            name: 'Resize',
            action: () => ResizeVolume.trigger(dispatch, record),
          },
          record.linode_id === null
            ? {
              name: 'Attach',
              action: () => AttachVolume.trigger(dispatch, linodes, record),
            }
            : {
              name: 'Detach',
              action: () => this.detachVolumes(record),
            },
        ],
      },
      { elements: [{ name: 'Delete', action: () => this.deleteVolumes(record) }] },
    ];

    return (
      <TableCell column={column} record={record} className="ActionsCell">
        <Dropdown
          groups={groups}
          analytics={{ title: 'Volume actions' }}
        />
      </TableCell>
    );
  }

  render() {
    const { dispatch, selectedMap, volumes, objectType, className } = this.props;
    const { filter } = this.state;

    const { sorted } = transform(volumes, {
      filterBy: filter,
    });

    return (
      <List>
        <ListHeader className="Menu">
          <div className="Menu-item">
            <MassEditControl
              data={sorted}
              dispatch={dispatch}
              massEditGroups={[{
                elements: [
                  { name: 'Delete', action: this.deleteVolumes },
                ],
              }]}
              selectedMap={selectedMap}
              objectType={objectType}
              toggleSelected={toggleSelected}
            />
          </div>
          <div className="Menu-item">
            <Input
              placeholder="Filter..."
              onChange={({ target: { value } }) => this.setState({ filter: value })}
              value={this.state.filter}
            />
          </div>
        </ListHeader>
        <ListBody>
          <Table
            columns={[
              { cellComponent: CheckboxCell, headerClassName: 'CheckboxColumn' },
              {
                cellComponent: LabelCell,
                headerClassName: 'LabelColumn',
                label: 'Label',
                dataKey: 'label',
                titleKey: 'label',
                tooltipEnabled: true,
              },
              {
                label: 'Size',
                dataFn: (volume) => {
                  const { size } = volume;
                  return `${size} GiB`;
                },
              },
              {
                cellComponent: RegionCell,
                headerClassName: 'RegionColumn',
                label: 'Region',
              },
              {
                cellComponent: this.renderAttached,
                className: 'AttachedCell',
                label: 'Attached',
              },
              {
                cellComponent: this.renderVolumeActions,
                headerClassName: 'ActionsColumn',
              },
            ]}
            data={sorted}
            selectedMap={selectedMap}
            disableHeader={!className}
            onToggleSelect={(record) => {
              dispatch(toggleSelected(objectType, record.id));
            }}
            className={className}
            noDataMessage="No Volumes found."
          />
        </ListBody>
      </List>
    );
  }
}

VolumesList.propTypes = {
  dispatch: PropTypes.func.isRequired,
  objectType: PropTypes.string.isRequired,
  selectedMap: PropTypes.object.isRequired,
  volumes: PropTypes.object.isRequired,
  className: PropTypes.string,
  linodes: PropTypes.object,
};
