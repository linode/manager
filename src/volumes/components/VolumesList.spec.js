import { shallow, mount } from 'enzyme';
import React from 'react';
import sinon from 'sinon';

import { SHOW_MODAL } from '~/actions/modal';
import VolumesList from './VolumesList';

import { api } from '~/data';

const { volumes: { volumes } } = api;

describe('linodes/volumes/components/VolumesList', () => {
  const sandbox = sinon.sandbox.create();

  afterEach(() => {
    sandbox.restore();
  });

  const dispatch = sandbox.spy();

  it.skip('renders a table', () => {
    const wrapper = shallow(
      <VolumesList
        dispatch={dispatch}
        selectedMap={{}}
        volumes={volumes}
        objectType="volumes"
      />
    );

    const vol = wrapper.find('Table');
    expect(vol.prop('data')).toEqual(volumes);
  });

  it.skip('shows the delete modal when delete is pressed', () => {
    const wrapper = mount(
      <VolumesList
        dispatch={dispatch}
        selectedMap={{}}
        volumes={volumes}
        objectType="volumes"
      />
    );
    const volDelete = wrapper.find('.TableRow Button').at(0);
    dispatch.reset();
    volDelete.simulate('click');
    expect(dispatch.callCount).toEqual(1);
    expect(dispatch.firstCall.args[0]).toHaveProperty('type', SHOW_MODAL);
  });

  it.skip('deletes selected volumes when delete is pressed', () => {
    const spy = jest.spyOn(VolumesList.prototype, 'deleteVolumes');
    const page = mount(
      <VolumesList
        dispatch={dispatch}
        selectedMap={{ 38: true }}
        volumes={volumes}
        objectType="volumes"
      />
    );

    const deleteButton = page.find('MassEditControl button#delete');
    deleteButton.simulate('click');

    expect(spy).toHaveBeenCalled();
  });
});
