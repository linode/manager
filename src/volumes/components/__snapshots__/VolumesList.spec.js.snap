// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linodes/volumes/components/VolumesList should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VolumesList
    dispatch={[MockFunction]}
    objectType="volumes"
    selectedMap={Object {}}
    volumes={
      Array [
        Object {
          "_polling": false,
          "created": "2017-08-08T13:55:16",
          "id": 38,
          "label": "test",
          "linode_id": null,
          "region": "us-east-1a",
          "size": 20,
          "status": "active",
          "updated": "2017-08-08T04:00:00",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ListHeader
          className="Menu"
        >
          <div
            className="Menu-item"
          >
            <MassEditControl
              data={
                Array [
                  Object {
                    "_polling": false,
                    "created": "2017-08-08T13:55:16",
                    "id": 38,
                    "label": "test",
                    "linode_id": null,
                    "region": "us-east-1a",
                    "size": 20,
                    "status": "active",
                    "updated": "2017-08-08T04:00:00",
                  },
                ]
              }
              dispatch={[MockFunction]}
              massEditGroups={
                Array [
                  Object {
                    "elements": Array [
                      Object {
                        "action": [Function],
                        "name": "Delete",
                      },
                    ],
                  },
                ]
              }
              objectType="volumes"
              selectedKey="id"
              selectedMap={Object {}}
              toggleSelected={[Function]}
            />
          </div>
          <div
            className="Menu-item"
          >
            <Input
              className=""
              onChange={[Function]}
              placeholder="Filter..."
              value=""
            />
          </div>
        </ListHeader>,
        <ListBody>
          <Table
            className=""
            columns={
              Array [
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "CheckboxColumn",
                },
                Object {
                  "cellComponent": [Function],
                  "dataKey": "label",
                  "headerClassName": "LabelColumn",
                  "label": "Label",
                  "titleKey": "label",
                  "tooltipEnabled": true,
                },
                Object {
                  "dataFn": [Function],
                  "label": "Size",
                },
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "RegionColumn",
                  "label": "Region",
                },
                Object {
                  "cellComponent": [Function],
                  "className": "AttachedCell",
                  "label": "Attached",
                },
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "ActionsColumn",
                },
              ]
            }
            data={
              Array [
                Object {
                  "_polling": false,
                  "created": "2017-08-08T13:55:16",
                  "id": 38,
                  "label": "test",
                  "linode_id": null,
                  "region": "us-east-1a",
                  "size": 20,
                  "status": "active",
                  "updated": "2017-08-08T04:00:00",
                },
              ]
            }
            disableHeader={true}
            noDataMessage="No Volumes found."
            onToggleSelect={[Function]}
            selectedMap={Object {}}
          />
        </ListBody>,
      ],
      "className": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <div
              className="Menu-item"
            >
              <MassEditControl
                data={
                  Array [
                    Object {
                      "_polling": false,
                      "created": "2017-08-08T13:55:16",
                      "id": 38,
                      "label": "test",
                      "linode_id": null,
                      "region": "us-east-1a",
                      "size": 20,
                      "status": "active",
                      "updated": "2017-08-08T04:00:00",
                    },
                  ]
                }
                dispatch={[MockFunction]}
                massEditGroups={
                  Array [
                    Object {
                      "elements": Array [
                        Object {
                          "action": [Function],
                          "name": "Delete",
                        },
                      ],
                    },
                  ]
                }
                objectType="volumes"
                selectedKey="id"
                selectedMap={Object {}}
                toggleSelected={[Function]}
              />
            </div>,
            <div
              className="Menu-item"
            >
              <Input
                className=""
                onChange={[Function]}
                placeholder="Filter..."
                value=""
              />
            </div>,
          ],
          "className": "Menu",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <MassEditControl
                data={
                  Array [
                    Object {
                      "_polling": false,
                      "created": "2017-08-08T13:55:16",
                      "id": 38,
                      "label": "test",
                      "linode_id": null,
                      "region": "us-east-1a",
                      "size": 20,
                      "status": "active",
                      "updated": "2017-08-08T04:00:00",
                    },
                  ]
                }
                dispatch={[MockFunction]}
                massEditGroups={
                  Array [
                    Object {
                      "elements": Array [
                        Object {
                          "action": [Function],
                          "name": "Delete",
                        },
                      ],
                    },
                  ]
                }
                objectType="volumes"
                selectedKey="id"
                selectedMap={Object {}}
                toggleSelected={[Function]}
              />,
              "className": "Menu-item",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "data": Array [
                  Object {
                    "_polling": false,
                    "created": "2017-08-08T13:55:16",
                    "id": 38,
                    "label": "test",
                    "linode_id": null,
                    "region": "us-east-1a",
                    "size": 20,
                    "status": "active",
                    "updated": "2017-08-08T04:00:00",
                  },
                ],
                "dispatch": [MockFunction],
                "massEditGroups": Array [
                  Object {
                    "elements": Array [
                      Object {
                        "action": [Function],
                        "name": "Delete",
                      },
                    ],
                  },
                ],
                "objectType": "volumes",
                "selectedKey": "id",
                "selectedMap": Object {},
                "toggleSelected": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Input
                className=""
                onChange={[Function]}
                placeholder="Filter..."
                value=""
              />,
              "className": "Menu-item",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "",
                "onChange": [Function],
                "placeholder": "Filter...",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Table
            className=""
            columns={
              Array [
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "CheckboxColumn",
                },
                Object {
                  "cellComponent": [Function],
                  "dataKey": "label",
                  "headerClassName": "LabelColumn",
                  "label": "Label",
                  "titleKey": "label",
                  "tooltipEnabled": true,
                },
                Object {
                  "dataFn": [Function],
                  "label": "Size",
                },
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "RegionColumn",
                  "label": "Region",
                },
                Object {
                  "cellComponent": [Function],
                  "className": "AttachedCell",
                  "label": "Attached",
                },
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "ActionsColumn",
                },
              ]
            }
            data={
              Array [
                Object {
                  "_polling": false,
                  "created": "2017-08-08T13:55:16",
                  "id": 38,
                  "label": "test",
                  "linode_id": null,
                  "region": "us-east-1a",
                  "size": 20,
                  "status": "active",
                  "updated": "2017-08-08T04:00:00",
                },
              ]
            }
            disableHeader={true}
            noDataMessage="No Volumes found."
            onToggleSelect={[Function]}
            selectedMap={Object {}}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "",
            "columns": Array [
              Object {
                "cellComponent": [Function],
                "headerClassName": "CheckboxColumn",
              },
              Object {
                "cellComponent": [Function],
                "dataKey": "label",
                "headerClassName": "LabelColumn",
                "label": "Label",
                "titleKey": "label",
                "tooltipEnabled": true,
              },
              Object {
                "dataFn": [Function],
                "label": "Size",
              },
              Object {
                "cellComponent": [Function],
                "headerClassName": "RegionColumn",
                "label": "Region",
              },
              Object {
                "cellComponent": [Function],
                "className": "AttachedCell",
                "label": "Attached",
              },
              Object {
                "cellComponent": [Function],
                "headerClassName": "ActionsColumn",
              },
            ],
            "data": Array [
              Object {
                "_polling": false,
                "created": "2017-08-08T13:55:16",
                "id": 38,
                "label": "test",
                "linode_id": null,
                "region": "us-east-1a",
                "size": 20,
                "status": "active",
                "updated": "2017-08-08T04:00:00",
              },
            ],
            "disableHeader": true,
            "noDataMessage": "No Volumes found.",
            "onToggleSelect": [Function],
            "selectedMap": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ListHeader
            className="Menu"
          >
            <div
              className="Menu-item"
            >
              <MassEditControl
                data={
                  Array [
                    Object {
                      "_polling": false,
                      "created": "2017-08-08T13:55:16",
                      "id": 38,
                      "label": "test",
                      "linode_id": null,
                      "region": "us-east-1a",
                      "size": 20,
                      "status": "active",
                      "updated": "2017-08-08T04:00:00",
                    },
                  ]
                }
                dispatch={[MockFunction]}
                massEditGroups={
                  Array [
                    Object {
                      "elements": Array [
                        Object {
                          "action": [Function],
                          "name": "Delete",
                        },
                      ],
                    },
                  ]
                }
                objectType="volumes"
                selectedKey="id"
                selectedMap={Object {}}
                toggleSelected={[Function]}
              />
            </div>
            <div
              className="Menu-item"
            >
              <Input
                className=""
                onChange={[Function]}
                placeholder="Filter..."
                value=""
              />
            </div>
          </ListHeader>,
          <ListBody>
            <Table
              className=""
              columns={
                Array [
                  Object {
                    "cellComponent": [Function],
                    "headerClassName": "CheckboxColumn",
                  },
                  Object {
                    "cellComponent": [Function],
                    "dataKey": "label",
                    "headerClassName": "LabelColumn",
                    "label": "Label",
                    "titleKey": "label",
                    "tooltipEnabled": true,
                  },
                  Object {
                    "dataFn": [Function],
                    "label": "Size",
                  },
                  Object {
                    "cellComponent": [Function],
                    "headerClassName": "RegionColumn",
                    "label": "Region",
                  },
                  Object {
                    "cellComponent": [Function],
                    "className": "AttachedCell",
                    "label": "Attached",
                  },
                  Object {
                    "cellComponent": [Function],
                    "headerClassName": "ActionsColumn",
                  },
                ]
              }
              data={
                Array [
                  Object {
                    "_polling": false,
                    "created": "2017-08-08T13:55:16",
                    "id": 38,
                    "label": "test",
                    "linode_id": null,
                    "region": "us-east-1a",
                    "size": 20,
                    "status": "active",
                    "updated": "2017-08-08T04:00:00",
                  },
                ]
              }
              disableHeader={true}
              noDataMessage="No Volumes found."
              onToggleSelect={[Function]}
              selectedMap={Object {}}
            />
          </ListBody>,
        ],
        "className": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <div
                className="Menu-item"
              >
                <MassEditControl
                  data={
                    Array [
                      Object {
                        "_polling": false,
                        "created": "2017-08-08T13:55:16",
                        "id": 38,
                        "label": "test",
                        "linode_id": null,
                        "region": "us-east-1a",
                        "size": 20,
                        "status": "active",
                        "updated": "2017-08-08T04:00:00",
                      },
                    ]
                  }
                  dispatch={[MockFunction]}
                  massEditGroups={
                    Array [
                      Object {
                        "elements": Array [
                          Object {
                            "action": [Function],
                            "name": "Delete",
                          },
                        ],
                      },
                    ]
                  }
                  objectType="volumes"
                  selectedKey="id"
                  selectedMap={Object {}}
                  toggleSelected={[Function]}
                />
              </div>,
              <div
                className="Menu-item"
              >
                <Input
                  className=""
                  onChange={[Function]}
                  placeholder="Filter..."
                  value=""
                />
              </div>,
            ],
            "className": "Menu",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <MassEditControl
                  data={
                    Array [
                      Object {
                        "_polling": false,
                        "created": "2017-08-08T13:55:16",
                        "id": 38,
                        "label": "test",
                        "linode_id": null,
                        "region": "us-east-1a",
                        "size": 20,
                        "status": "active",
                        "updated": "2017-08-08T04:00:00",
                      },
                    ]
                  }
                  dispatch={[MockFunction]}
                  massEditGroups={
                    Array [
                      Object {
                        "elements": Array [
                          Object {
                            "action": [Function],
                            "name": "Delete",
                          },
                        ],
                      },
                    ]
                  }
                  objectType="volumes"
                  selectedKey="id"
                  selectedMap={Object {}}
                  toggleSelected={[Function]}
                />,
                "className": "Menu-item",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "data": Array [
                    Object {
                      "_polling": false,
                      "created": "2017-08-08T13:55:16",
                      "id": 38,
                      "label": "test",
                      "linode_id": null,
                      "region": "us-east-1a",
                      "size": 20,
                      "status": "active",
                      "updated": "2017-08-08T04:00:00",
                    },
                  ],
                  "dispatch": [MockFunction],
                  "massEditGroups": Array [
                    Object {
                      "elements": Array [
                        Object {
                          "action": [Function],
                          "name": "Delete",
                        },
                      ],
                    },
                  ],
                  "objectType": "volumes",
                  "selectedKey": "id",
                  "selectedMap": Object {},
                  "toggleSelected": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Input
                  className=""
                  onChange={[Function]}
                  placeholder="Filter..."
                  value=""
                />,
                "className": "Menu-item",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "",
                  "onChange": [Function],
                  "placeholder": "Filter...",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Table
              className=""
              columns={
                Array [
                  Object {
                    "cellComponent": [Function],
                    "headerClassName": "CheckboxColumn",
                  },
                  Object {
                    "cellComponent": [Function],
                    "dataKey": "label",
                    "headerClassName": "LabelColumn",
                    "label": "Label",
                    "titleKey": "label",
                    "tooltipEnabled": true,
                  },
                  Object {
                    "dataFn": [Function],
                    "label": "Size",
                  },
                  Object {
                    "cellComponent": [Function],
                    "headerClassName": "RegionColumn",
                    "label": "Region",
                  },
                  Object {
                    "cellComponent": [Function],
                    "className": "AttachedCell",
                    "label": "Attached",
                  },
                  Object {
                    "cellComponent": [Function],
                    "headerClassName": "ActionsColumn",
                  },
                ]
              }
              data={
                Array [
                  Object {
                    "_polling": false,
                    "created": "2017-08-08T13:55:16",
                    "id": 38,
                    "label": "test",
                    "linode_id": null,
                    "region": "us-east-1a",
                    "size": 20,
                    "status": "active",
                    "updated": "2017-08-08T04:00:00",
                  },
                ]
              }
              disableHeader={true}
              noDataMessage="No Volumes found."
              onToggleSelect={[Function]}
              selectedMap={Object {}}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "",
              "columns": Array [
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "CheckboxColumn",
                },
                Object {
                  "cellComponent": [Function],
                  "dataKey": "label",
                  "headerClassName": "LabelColumn",
                  "label": "Label",
                  "titleKey": "label",
                  "tooltipEnabled": true,
                },
                Object {
                  "dataFn": [Function],
                  "label": "Size",
                },
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "RegionColumn",
                  "label": "Region",
                },
                Object {
                  "cellComponent": [Function],
                  "className": "AttachedCell",
                  "label": "Attached",
                },
                Object {
                  "cellComponent": [Function],
                  "headerClassName": "ActionsColumn",
                },
              ],
              "data": Array [
                Object {
                  "_polling": false,
                  "created": "2017-08-08T13:55:16",
                  "id": 38,
                  "label": "test",
                  "linode_id": null,
                  "region": "us-east-1a",
                  "size": 20,
                  "status": "active",
                  "updated": "2017-08-08T04:00:00",
                },
              ],
              "disableHeader": true,
              "noDataMessage": "No Volumes found.",
              "onToggleSelect": [Function],
              "selectedMap": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
